### PR TITLE
COSE Signature Envelope

### LINKED ISSUE #117: Implement COSE as signing envelope
I'd like to see Notary v2 support [CBOR Object Signing and Encryption (COSE)](https://datatracker.ietf.org/doc/html/rfc8152) as the signing envelope for content, as it is flexible to support all of the Nv2 signing requirements including [IETF draft specs on support for x.509](https://datatracker.ietf.org/doc/html/draft-ietf-cose-x509-08). 

Also, COSE has broader alignments across the industry in CoMID/CoSWID, and standards for software manifests and remote attestation including C2PA .

Consider this as an update to [Issue #99 ](https://github.com/notaryproject/notaryproject/issues/99).

### PR DESCRIPTION
**Pull request by shizhMSFT**
Replace JWS with COSE for the envelope format.

Highlight: Unlike JWS, `alg` is not a required header in COSE. Therefore, we can remove it completely, making the signature envelope more secure.

Resolves #117

### CODE REVIEW

File: signature-specification.md
Comment (briankr-ms) [inline signature-specification.md]: I would suggest clarifying that other headers can be included, but Notary v2 will not process them (and will fail if they are set as 'crit'). For example, some COSE implementations will always add 'alg' - it's fine if nv2 won't rely on this, but I don't believe it should be a failure if it is present.

File: signature-specification.md
Comment (briankr-ms) [inline signature-specification.md]: Clarify that these are not just "integers", but tagged [CBOR date/time values](https://datatracker.ietf.org/doc/html/rfc7049#section-2.4.1). Consider supporting other time formats such as RFC3339 or future [CBOR time](https://datatracker.ietf.org/doc/html/draft-ietf-cbor-time-tag).

File: signature-specification.md
Comment (briankr-ms) [inline signature-specification.md:164]: There may be value in including x5chain in the protected headers to ensure certs cannot be substituted in an attack. If not (or in addition to) that, x5t/34 should be a protected header.

File: signature-specification.md
Comment (briankr-ms) [inline signature-specification.md:174]: If x5chain is not protected, should require that x5t/34 is. There may also be value in simply requiring that x5chain is protected. This guards against a potential cert substitution attack, at the expense of requiring a larger signature envelope size because the header can't be stripped in environments where the value is known externally.

File: signature-specification.md
Comment (briankr-ms) [inline signature-specification.md:216]: If detached payloads are not supported, I suggest documenting that explicitly in this spec.

**Comment by roywill** [comment]
>   - **`org.cncf.notary.x509certs.fingerprint.sha256`**: A REQUIRED annotation whose value contains the list of SHA-256 fingerprint of signing certificate and certificate chain used for signature generation.

If we adopt\moving to a Ledger type system, why would we keep the certificate chain anywhere?  Is this a hold over from threat?

---
Refers to: signature-specification.md:25 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

**Comment by roywill** [comment]
>             "mediaType": "application/jose+json",

Why Jose?

---
Refers to: signature-specification.md:33 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

**Comment by roywill** [comment]
> The Notary v2 signature artifact's `org.cncf.notary.x509certs.fingerprint.sha256` annotations key MUST contain the list of SHA-256 fingerprints of certificate and certificate chain used for signing.

Why do we need this?

---
Refers to: signature-specification.md:59 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

**Comment by roywill** [comment]
> - Signed attributes `v`: The signature metadata that is integrity protected - e.g. signature expiration time, signing time, etc.

Why do we keep the expiration time in its own field.  Is this part of the support for extending life based on Secure TimeStamp data?

---
Refers to: signature-specification.md:67 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

**Comment by roywill** [comment]
> - Unsigned attributes `u`: This OPTIONAL property represents signature metadata that is not integrity protected - e.g. timestamp, certificates, etc.

In the case of our other scenarios, the key thing that we store in unprotected headers is another signed message.  How much of this is true here?

---
Refers to: signature-specification.md:68 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

---
Edit from @SteveLasker 

Updated to reflect:

- Unsigned attributes `u`: This OPTIONAL property represents signature metadata that is not integrity protected by this envelope- e.g. timestamp, certificates, etc.
  We anticipate all unsigned attributes contain signed content.

**Comment by roywill** [comment]
>       "io.wabbit-networks.buildId": "123"  // user defined signed attribute.

It would be nice to show something under io.cncf.notary

---
Refers to: signature-specification.md:91 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

--- 
@SteveLasker  Commenting in an edit:

We could switch to a CNCF project. Within the notary project, we use two companies as representative users
- Wabbit Networks is a small ISV, shipping their software
- Acme Rockets is a small consumer, that consumes public software, such as the `net-monitor:v1` image from Wabbit Networks. 
This is just to be consistent with those users.

**Comment by roywill** [comment]
>    "mediaType": "sbom/example",

Media type is usually text\plain, multipart\related, image\jpg .

Did we really define a type sbom/example?

---
Refers to: signature-specification.md:98 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

---
Inlined edit from @SteveLasker 

To avoid referencing specific projects, we use `[artifactType]/example`

**Comment by roywill** [comment]
>   Its value should be numeric representing the number of seconds (not milliseconds) since Epoch.

Is it the time it was signed or the time the payload was created?  Thus in a disaster scenario when things come back up is it the time the "operation happened" or the time when we requested a signature?

---
Refers to: signature-specification.md:109 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

**Comment by roywill** [comment]
>   Its value should be numeric representing the number of seconds since Epoch.

Isn't this also expected to be > than Creation Time?  Or is it truly free form and thus can be earlier than or equal to creation time?

---
Refers to: signature-specification.md:111 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

**Comment by roywill** [comment]
> - **Certificates**: The ordered collection of X.509 certificates with a signing certificate as the first entry.

Why are all the intermediaries here and does it include the root?
What happens if we have to support multiple signatures?

---
Refers to: signature-specification.md:116 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

**Comment by roywill** [comment]
> - **Timestamp**:  The time stamp token generated for a given signature.

This as it stands requires a CMS parser to exist to understand it.  The IETF has work on its place to come up with another representation and as such is this something we can postpone?  

---
Refers to: signature-specification.md:117 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

**Comment by roywill** [comment]
> - Signed attributes `v`: The signature metadata that is integrity protected - e.g. signature expiration time, signing time, etc.

The signature was valid at the time pen went to paper.  Really what you are alluding to is when support stops or after which it is impossible to check revocation.  Which is it?

---
In reply to: [1067417710](https://github.com/notaryproject/notaryproject/pull/139#issuecomment-1067417710) [](http://example.com/codeflow?ancestors=1067417710,1067417710)

---
Refers to: signature-specification.md:67 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

**Comment by briankr-ms** [comment]
> > * **Timestamp**:  The time stamp token generated for a given signature.
> 
> This as it stands requires a CMS parser to exist to understand it. The IETF has work on its place to come up with another representation and as such is this something we can postpone?
> 
> Refers to: signature-specification.md:117 in [644c8cd](https://github.com/notaryproject/notaryproject/commit/644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2). [](commit_id = [644c8cd](https://github.com/notaryproject/notaryproject/commit/644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2), deletion_comment = False)

An RFC3161 countersignature is specifically useful as independent evidence that a signature was created (or at least existed) within the validity period of an x.509 certificate, for the purpose of determining whether it can be considered valid after the signing certificate's expiration.

I think this is distinct from other time-related claims (i.e. "signingtime" or "expiry") that are headers protected by the same key that signed the payload, and which are potentially meaningful in a non-x.509 trust model where key and signature validity are not associated with a certificate. So while there is some appearance of redundancy here, ultimately they serve different purposes.

Given that x.509 already implies the ASN.1 overhead (if not CMS), and that RFC3161 countersignatures are already commonly used to establish validity after a certificate's expiration, I think it's reasonable to use the existing standard for this purpose, and that this is preferable to omitting it until a CBOR variant is standardized/widely available. Without it, you run into scenarios where an artifact is signed shortly before a certificate's expiration and is only "valid" for a brief period. If and when the trust model moves away from x.509, this field is likely no longer needed or useful.

**Comment by dlorenc** [comment]
> If we adopt\moving to a Ledger type system, why would we keep the certificate chain anywhere? Is this a hold over from threat?

Is there any documentation on the Ledger system?

**Comment by kaywilliams** [comment]
> Is there any documentation on the Ledger system?

The reference is to an early proposal in the IETF. 

[https://datatracker.ietf.org/doc/draft-birkholz-scitt-architecture/](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdatatracker.ietf.org%2Fdoc%2Fdraft-birkholz-scitt-architecture%2F&data=04%7C01%7Ckayw%40microsoft.com%7Cb4a5861e8f004178495008da06b83947%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637829685549522253%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=f1TerknB0Jo%2FEm%2FK3lQRivb9xbdCSyNMVPG%2B6SP41Us%3D&reserved=0)
[https://datatracker.ietf.org/doc/draft-birkholz-scitt-receipts/](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdatatracker.ietf.org%2Fdoc%2Fdraft-birkholz-scitt-receipts%2F&data=04%7C01%7Ckayw%40microsoft.com%7Cb4a5861e8f004178495008da06b83947%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637829685549572227%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Ts9Y3glmJZKdiO141BRit58tUu1ZYSRhUqbRrrMJQFE%3D&reserved=0)
https://github.com/ietf-scitt

Discussion is happening on the ietf-scitt mailing list:
[SCITT Info Page (ietf.org)](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.ietf.org%2Fmailman%2Flistinfo%2Fscitt&data=04%7C01%7Ckayw%40microsoft.com%7Cb4a5861e8f004178495008da06b83947%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637829685549522253%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=nL2GfNrpY6mqBrqRVqe5jcKgYkJJ2rRZybVN%2FRyoXW8%3D&reserved=0)

File: signature-specification.md
Comment (shizhMSFT) [reply signature-specification.md:174]: @priteshbandi Should we move `Certificates` attribute from _Unsigned Attributes_ to _Signed Attributes_?

**Comment by shizhMSFT** [comment]
@priteshbandi @SteveLasker Could you also take a look at @roywill 's comment since his comment is mainly for the `signature specification` itself and is not specific to this PR?

**Comment by shizhMSFT** [comment]
>             "mediaType": "application/jose+json",

updated to `cose`.

---
In reply to: [1067416112](https://github.com/notaryproject/notaryproject/pull/139#issuecomment-1067416112) [](http://example.com/codeflow?ancestors=1067416112,1067416112)

---
Refers to: signature-specification.md:33 in 644c8cd. [](commit_id = 644c8cdcc702be5f18f2e5c0ee8ae9d5486cf0c2, deletion_comment = False)

**Comment by letmaik** [comment]
> Highlight: Unlike JWS, `alg` is not a required header in COSE. Therefore, we can remove it completely, making the signature envelope more secure.

Keeping `alg` in the header is useful when relying on key discovery systems that don't bind the alg to the key itself. An example is DID where `alg` can be included in the JWK object. Since different DID methods have more or less guarantees on the integrity of the resolved DID document (DID documents are unsigned), e.g. when using ledgers vs. fetching a URL, it makes sense to reduce reliance on additional metadata as much as possible.

**Comment by shizhMSFT** [comment]
> > Highlight: Unlike JWS, `alg` is not a required header in COSE. Therefore, we can remove it completely, making the signature envelope more secure.
> 
> Keeping `alg` in the header is useful when relying on key discovery systems that don't bind the alg to the key itself. An example is DID where `alg` can be included in the JWK object. Since different DID methods have more or less guarantees on the integrity of the resolved DID document (DID documents are unsigned), e.g. when using ledgers vs. fetching a URL, it makes sense to reduce reliance on additional metadata as much as possible.

The `alg` header is OPTIONAL and many COSE implementations still set values to it. Unlike JWS, the verification won't fail if `alg` does not present.

**Comment by SteveLasker** [comment]
Proposing we limit the scope of this effort to `COSE_SIGN1`, deferring `COSE_SIGN` to a future date.

File: signature-specification.md
Comment (letmaik) [inline signature-specification.md:153]: Can you add **`iss`** (*string*) to this list as optional property indicating the issuer of the signature? If it follows a well-known scheme like DID (starting with `did:`) then it may be used for key discovery, together with `kid` which is also missing here.

Comment (shizhMSFT) [reply signature-specification.md:153]: Currently, Notary v2 only supports certs. Other key discovery systems like `kid` or `iss` or others are not supported, and are not covered in this PR. This PR focuses on changing what we have for JWS to COSE and nothing else. BTW, `iss` makes it more like a CWT.

@letmaik Could you follow up #95 or open a new issue in case we need to support other key discovery systems so that the community can review?

**Comment by SteveLasker** [comment]
@shizhMSFT, you’re correct this PR is focused on COSE and x509. And, we’re looking to use DiD for where to discover the public key. Knowing we’ll add other key and identity types, is there anything we would add now to make that support less of a breaking change?

**Comment by SteveLasker** [comment]
From today's Notary call, we agreed to keep both the current JWT, adding COSE. 

We’re hoping to land the COSE format for the initial release, but we’ve said notary will support multiple signature formats over time, for instance, different governments (China has been the common example, but also support for newer fomats over time). If we add multiples now, we validate the spec supports multiples. If we land COSE for v1, we can pull JWT out, knowing the spec supports multiple formats.

**Comment by SteveLasker** [comment]
To move the spec forward, aleviating Shiwei to focus on the go-cose work, suggesting we merge this into [cose-envelope](https://github.com/notaryproject/notaryproject/tree/cose-envelope) branch.
Please LGTM as the PR target was changed.

**Review by SteveLasker — APPROVED** [review]
LGTM to merge into the cose-envelope branch, and we'll continue to iterate the feedback there.

**Comment by gokarnm** [comment]
LGTM to merge into the cose-envelope branch, to unblock work. @priteshbandi  and me have yet to review this PR completely.

**Review by gokarnm — APPROVED** [review]

File: signature-specification.md
Comment (SteveLasker) [reply signature-specification.md:174]: @briankr-ms , @shizhMSFT what's the call here? I can make changes if we can conclude on a decision.

**Comment by SteveLasker** [comment]
> > The Notary v2 signature artifact's `org.cncf.notary.x509certs.fingerprint.sha256` annotations key MUST contain the list of SHA-256 fingerprints of certificate and certificate chain used for signing.
> 
> Why do we need this?
> 

@shizhMSFT, @roywill, @briankr-ms what's the final call here?
Do we need an issue to track this?

**Comment by SteveLasker** [comment]
> Signed attributes v: The signature metadata that is integrity protected - e.g. signature expiration time, signing time, etc.

> Why do we keep the expiration time in its own field. Is this part of the support for extending life based on Secure TimeStamp data?

Need a call here: @roywill, @shizhMSFT, @gokarnm

**Comment by SteveLasker** [comment]
> Its value should be numeric representing the number of seconds (not milliseconds) since Epoch.

> Is it the time it was signed or the time the payload was created? Thus in a disaster scenario when things come back up is it the time the "operation happened" or the time when we requested a signature?

@shizhMSFT, @gokarnm, @priteshbandi, what would you like to make this?

**Comment by SteveLasker** [comment]
> Its value should be numeric representing the number of seconds since Epoch.

> Isn't this also expected to be > than Creation Time? Or is it truly free form and thus can be earlier than or equal to creation time?

@shizhMSFT, @gokarnm, @priteshbandi, same question here ^

**Comment by SteveLasker** [comment]
@roywill 
> Certificates: The ordered collection of X.509 certificates with a signing certificate as the first entry.
> Why are all the intermediaries here and does it include the root?
> What happens if we have to support multiple signatures?

Notary doesn't support multiple signatures in the same payload.  You can create multiple detached signatures, each with their own `subject` and each with their own claim. 

However, this does call out "signing certificate", which we've since said is not required. Should this be updated to say root certificate as the first entry?

File: signature-specification.md
Comment (gokarnm) [inline signature-specification.md:20]: Is the intent to disregard any optional parameter e.g. `application/cose; cose-type="cose-sign1"` ? Notation should probably include `code-type` param to be more specific.

File: signature-specification.md
Comment (gokarnm) [inline signature-specification.md:138]: Is there a reason to use headers of type string instead of int? Do we plan to followup and include these in IANA?

https://www.iana.org/assignments/cose/cose.xhtml#header-parameters

File: signature-specification.md
Comment (gokarnm) [reply signature-specification.md:174]: @briankr-ms could you provide more details for cert substitution attack. Notation will use x5chain in conjunction with trust store configured by the user. As part of verification, the cert chain is checked to terminate in one of the certs in trust store. @shizhMSFT  are we using x5t/thumbprint in any other context, other than signature manifest attributes, for matching signature to policy for evaluation?

Comment (gokarnm) [reply signature-specification.md:174]: @shizhMSFT The certs in chain are already signed by the next cert in the chain, and is expected to be authenticated against the trust store. Open to discussing if moving it to signed attributes protects us against certain attacks.

File: signature-specification.md
Comment (gokarnm) [inline signature-specification.md:132]: I'd explicitly mention that `alg` if included will not be considered by Notation during verification.

**Comment by gokarnm** [comment]
> > > The Notary v2 signature artifact's `org.cncf.notary.x509certs.fingerprint.sha256` annotations key MUST contain the list of SHA-256 fingerprints of certificate and certificate chain used for signing.
> > 
> > 
> > Why do we need this?
> 
> @shizhMSFT, @roywill, @briankr-ms what's the final call here? Do we need an issue to track this?

Multiple signatures can be associated with and OCI artifact, which are can be queried through the ORAS references API. This field is used to filter down signatures that will be considered for signature verification against the customer configured Trust store.

More context in #131

**Comment by gokarnm** [comment]
> > Signed attributes v: The signature metadata that is integrity protected - e.g. signature expiration time, signing time, etc.
> 
> > Why do we keep the expiration time in its own field. Is this part of the support for extending life based on Secure TimeStamp data?
> 
> Need a call here: @roywill, @shizhMSFT, @gokarnm

The signature can be generated by ephemeral keys, the optional signature expiry field allows the signatures to be valid beyond key expiry without relying on TSA.

Can be renamed to signature-expiry. Clarify that signature-expiry can be shorter or larger than key expiry.

File: signature-specification.md
Comment (gokarnm) [reply signature-specification.md:132]: TODO: Capture the intent for x5chain.

**Pull request merged**
